### PR TITLE
Fix for iterations space with automated model-data promotion

### PR DESCRIPTION
**Pull request by mlange05**
## Status

**REVIEW**

## Description
When using the new `!$claw model data` API for marking promotion variables in combination with a model-config file, the insertion of new loops in the CPU and GPU backends ignores any explicitly provided `[dimensions.iteration]` variables. This PR fixes this and adds a new test to demonstrate the correct insert of CPU and GPU loops over iteration ranges that are different to the provided size variables. In this test the loops inserted iteration over the range `DO p = pstart , pend, 1`, while the promoted variable dimension is `t ( 1 : nproma, nz)`.

### CODE REVIEW

**Comment by lxavier** [comment]
@mlange05 seems when of the test got stuck. Is this PR ready for review ?

**Comment by mlange05** [comment]
@lxavier From my end, yes I think it is ready for review. I don't have access to the travis-ci for the PR, but locally, the claw-sca tests ran fine for me.

File: cx2t/src/claw/tatsu/xcodeml/abstraction/DimensionDefinition.java [hunk 1]
```diff
@@ -228,6 +228,27 @@ public Xnode generateIndexRange(XcodeML xcodeml, boolean withStep) {
...
+  /**
+   * Generate the correct iterationRange element with iterationLowerBound,
+   * iterationUpperBound and step from the current dimension.
+   *
+   * @param xcodeml  Current XcodeML program unit in which elements will be
+   *                 created.
+   * @param withStep IF true, step element is created.
+   * @return A new indexRange elements.
+   */
+  public Xnode generateIterationRange(XcodeML xcodeml, boolean withStep) {
```
Comment (clementval) [inline cx2t/src/claw/tatsu/xcodeml/abstraction/DimensionDefinition.java:240]: This method seems very similar to the one on line 219. Might be a good idea to try to combined them if possible. Or having a base method where we pass the lower and upper bound. 

```java
public Xnode generateIndexRange(XcodeML xcodeml, boolean withStep) {
```

**Review by clementval — APPROVED** [review]
Just a small comment

**Comment by FrostyMike** [comment]
Travis tests are green. This was a glitch in github/travis integration, sometimes it does not update test results.

**Review by FrostyMike — APPROVED** [review]
Only checked that there will be no conflicts with upcoming build-change PRs.

File: cx2t/src/claw/tatsu/xcodeml/abstraction/DimensionDefinition.java (hunk 1 above)
Comment (mlange05) [reply cx2t/src/claw/tatsu/xcodeml/abstraction/DimensionDefinition.java]: @clementval Yes, I agree, we can merge the two and use a flag to separate between size and iteration indices. I pushed the according change.

Comment (FrostyMike) [reply cx2t/src/claw/tatsu/xcodeml/abstraction/DimensionDefinition.java]: Please fix the unit tests. Their compilation fails after latest change.

Comment (mlange05) [reply cx2t/src/claw/tatsu/xcodeml/abstraction/DimensionDefinition.java]: Yes, sorry, had missed that. Fixed now.

**Pull request merged**